File: DB/src/main/java/adapter/converter/ShelfProto/ConverterShelf.java
package adapter.converter.ShelfProto;

import Domain.Model.Item;
import Domain.Model.ItemType;
import Domain.Model.Shelf;
import Domain.Model.User;
import GRPC.proto.File.*;
import adapter.converter.itemProto.ConverterItem;
import adapter.converter.itemTypeProto.ConverterItemType;
import adapter.converter.userProto.ConverterUser;

import java.util.ArrayList;

public class ConverterShelf
{

  public static ShelfProto shelf_to_ShelfProto(Shelf shelf)
  {
    ArrayList<ItemProto> itemProtos = new ArrayList<>();

    for (Item item : shelf.getItemsOnShelf())
    {
      itemProtos.add(ItemProto.newBuilder()
          .setType(ConverterItemType.itemType_To_ItemTypeProto(item.getType()))
          .setUniqueID(item.getUid())
          .setOwner(ConverterUser.User_To_UserProto(item.getOwner()))
          .build());


    }
    ShelfProto result=ShelfProto.newBuilder()
            .setRowNo(shelf.getRowNo())
            .setShelfNo(shelf.getShelfNo())
            .setShelfDimX(shelf.getDimX())
            .setShelfDimY(shelf.getDimY())
            .setShelfDimZ(shelf.getDimZ())
            .build();


    return result;
  }



  public static Shelf shelfProto_To_shelf(ShelfProto shelfproto)
  {
    ArrayList<Item> items = new ArrayList<>();

    Shelf shelf = new Shelf(shelfproto.getRowNo(),
        shelfproto.getShelfNo(),
        shelfproto.getShelfDimX(),
        shelfproto.getShelfDimY(),
        shelfproto.getShelfDimZ(),
        items);

    for (ItemProto item : shelfproto.getItemsOnShelfList())
    {
      ItemType itemType = ConverterItemType.itemTypeProto_To_ItemType(item.getType());
      User user = ConverterUser.UserProto_To_User(item.getOwner());

      items.add(new Item(itemType, item.getUniqueID(),user, shelf));
    }
    return shelf;
  }

}

File: Logic/AdapterToGRPC/Converter/ConverterShelf.cs
using GRPC.Proto;
using Shared.Model;

namespace Logic.AdapterToGRPC;

public class ConverterShelf
{
    public static Shared.Model.Shelf ShelfProtoToShelf(ShelfProto shelfProto){
        
    Shared.Model.Shelf shelf = new Shared.Model.Shelf();
    shelf.DimX = shelfProto.DimX;
    shelf.DimY = shelfProto.Dimy;
    shelf.DimZ = shelfProto.Dimz;
    shelf.RowNo = shelfProto.RowNo;
    shelf.ShelfNo = shelfProto.ShelfNo;
            
    shelf.ItemsOnShelf = new List<Shared.Model.Item>();
    foreach (var itemss in shelfProto.ItemsOnShelf)
    {

        User userit = ConverterUser.UserProtoToUser(itemss.Owner);

        itemType _itemType = new itemType(itemss.Type.Id, itemss.Type.DimX, itemss.Type.DimY, itemss.Type.DimZ);
                
        Shared.Model.Item item = new Shared.Model.Item(_itemType, itemss.UniqueID, userit, shelf);
        shelf.ItemsOnShelf.Add(item);
    }

    return shelf;
    }

    public static ShelfProto ShelfToShelfProto(Shared.Model.Shelf shelf)
    {
        return new ShelfProto
        {
            Dimy = shelf.DimY, Dimz = shelf.DimZ, DimX = shelf.DimX, RowNo = shelf.RowNo, ShelfNo = shelf.ShelfNo
        };
    }

    public static List<Shared.Model.Shelf> AllShelvesProtoToAllShelves(ShelvesListProto shelvesListProto)
    {
        List<Shared.Model.Shelf> result = new List<Shared.Model.Shelf>();
        
        foreach (ShelfProto protoList in shelvesListProto.Proto)
        {
            result.Add(ShelfProtoToShelf(protoList));
        }

        return result;
    }
}